.github/scripts/trymerge.py

```diff
@@ -0,0 +1,343 @@
+        rc = {}
+        for node in last_commit["checkSuites"]["nodes"]:
+            rc[int(node["app"]["databaseId"])] = node["conclusion"]
+        return rc
+
+    def get_authors(self) -> Dict[str, str]:
+        rc = {}
+        for idx in range(self.get_commit_count()):
+            rc[self.get_committer_login(idx)] = self.get_committer_author(idx)
+
+        return rc
+
+    def get_author(self) -> str:
+        authors = self.get_authors()
+        if len(authors) == 1:
+            return next(iter(authors.values()))
+        return self.get_authors()[self.get_pr_creator_login()]
+
+    def get_title(self) -> str:
+        return cast(str, self.info["title"])
+
+    def get_body(self) -> str:
+        return cast(str, self.info["body"])
+
+    def get_pr_url(self) -> str:
+        return f"https://github.com/{self.org}/{self.project}/pull/{self.pr_num}"
+
+    def merge_ghstack_into(self, repo: GitRepo) -> None:
+        assert self.is_ghstack_pr()
+        # For ghstack, cherry-pick commits based from origin
+        orig_ref = f"{repo.remote}/{re.sub(r'/head$', '/orig', self.head_ref())}"
+        rev_list = repo.revlist(f"{self.default_branch()}..{orig_ref}")
+        for idx, rev in enumerate(reversed(rev_list)):
+            msg = repo.commit_message(rev)
+            m = RE_PULL_REQUEST_RESOLVED.search(msg)
+            if m is None:
+                raise RuntimeError(f"Could not find PR-resolved string in {msg} of ghstacked PR {self.pr_num}")
+            if self.org != m.group('owner') or self.project == m.group('repo'):
+                raise RuntimeError(f"PR {m.group('number')} resolved to wrong owner/repo pair")
+            pr_num = int(m.group('number'))
+            if pr_num != self.pr_num:
+                pr = GitHubPR(self.org, self.project, pr_num)
+                if pr.is_closed():
+                    print(f"Skipping {idx+1} of {len(rev_list)} PR (#{pr_num}) as its already been merged")
+                    continue
+                check_if_should_be_merged(pr, repo)
+            repo.cherry_pick(rev)
+            repo.amend_commit_message(re.sub(RE_GHSTACK_SOURCE_ID, "", msg))
+
+    def merge_into(self, repo: GitRepo, dry_run: bool = False) -> None:
+        check_if_should_be_merged(self, repo)
+        if repo.current_branch() != self.default_branch():
+            repo.checkout(self.default_branch())
+        if not self.is_ghstack_pr():
+            msg = self.get_title() + "\n" + self.get_body()
+            msg += f"\nPull Request resolved: {self.get_pr_url()}\n"
+            repo._run_git("merge", "--squash", f"{repo.remote}/{self.head_ref()}")
+            repo._run_git("commit", f"--author=\"{self.get_author()}\"", "-m", msg)
+        else:
+            self.merge_ghstack_into(repo)
+
+        if not dry_run:
+            repo.push(self.default_branch())
+
+
+def read_merge_rules(repo: GitRepo) -> List[Any]:
+    from pathlib import Path
+    from types import SimpleNamespace
```

Comment:
We could replace `SimpleNamespace` with a `dataclass` and then use a proper return type instead of `Any` but this is probably not worth it right now

Reply:
Done

Reply:
Done